pkg/kubelet/kubelet_test.go

```diff
@@ -2154,6 +2156,65 @@ func TestMakeEnvironmentVariables(t *testing.T) {
 				"KUBERNETES_RO_PORT_8087_TCP_ADDR=1.2.3.7"),
 			21,
 		},
+		{
+			"simple substitution",
+			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name: "test-pod"}},
+			&api.Container{
+				Env: []api.EnvVar{
+					{Name: "BAR", Value: "bar"},
+					{Name: "FOO", Value: "${BAR}"},
+				},
+			},
+			"kubernetes",
+			true,
+			util.NewStringSet(
+				"BAR=bar",
+				"FOO=bar",
+			),
+			2,
+		},
+		{
+			"slightly more complex substitution",
+			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name: "test-pod"}},
+			&api.Container{
+				Env: []api.EnvVar{
+					{Name: "BAR", Value: "bar"},
```

Comment:
We could introduce a separate field on EnvVar that is the source of the data, and disallow value via validation.  That would clearly delineate the two.

----- Original Message -----
> > +				},
> > +			},
> > +			"kubernetes",
> > +			true,
> > +			util.NewStringSet(
> > +				"BAR=bar",
> > +				"FOO=bar",
> > +			),
> > +			2,
> > +		},
> > +		{
> > +			"slightly more complex substitution",
> > +			&api.Pod{ObjectMeta: api.ObjectMeta{Namespace: "unknown", Name:
> > "test-pod"}},
> > +			&api.Container{
> > +				Env: []api.EnvVar{
> > +					{Name: "BAR", Value: "bar"},
> 
> What will Docker's transitive expansion do for us? I looked at the code, but
> have forgotten already. I'd rather define a simpler mechanism to populate
> environment variables with Kubernetes data, and then rely upon Docker and/or
> the shell for more general expansion/substitution.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/6000/files#r27256435